From 7ef3eb6ee786d5a08eabdd1e6788fae9ab5ff8a4 Mon Sep 17 00:00:00 2001
From: Jan Petrous <jan.petrous@nxp.com>
Date: Mon, 20 Feb 2023 10:19:47 +0100
Subject: [PATCH 57/63] net: stmmac: Don't allow systime modifications for
 external systime

When MAC is configured to external systime mode, the adjustment
or setting of systime has no effect. Same for modifying the clock
frequency. Report EPERM back to the caller in this case.

Issue: ALB-9754
Upstream-Status: Pending 

Signed-off-by: Jan Petrous <jan.petrous@nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 drivers/net/ethernet/stmicro/stmmac/stmmac_ptp.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/drivers/net/ethernet/stmicro/stmmac/stmmac_ptp.c b/drivers/net/ethernet/stmicro/stmmac/stmmac_ptp.c
index 487418ef9b4f..7f153e3f200e 100644
--- a/drivers/net/ethernet/stmicro/stmmac/stmmac_ptp.c
+++ b/drivers/net/ethernet/stmicro/stmmac/stmmac_ptp.c
@@ -28,6 +28,10 @@ static int stmmac_adjust_freq(struct ptp_clock_info *ptp, s32 ppb)
 	int neg_adj = 0;
 	u64 adj;
 
+	/* Adjusting freq not possible when using external TS clock. */
+	if (priv->plat->ext_sys_time)
+		return -EPERM;
+
 	if (ppb < 0) {
 		neg_adj = 1;
 		ppb = -ppb;
@@ -65,6 +69,10 @@ static int stmmac_adjust_time(struct ptp_clock_info *ptp, s64 delta)
 	bool xmac, est_rst = false;
 	int ret;
 
+	/* Adjusting systime not possible when using external TS clock. */
+	if (priv->plat->ext_sys_time)
+		return -EPERM;
+
 	xmac = priv->plat->has_gmac4 || priv->plat->has_xgmac;
 
 	if (delta < 0) {
@@ -162,6 +170,10 @@ static int stmmac_set_time(struct ptp_clock_info *ptp,
 	    container_of(ptp, struct stmmac_priv, ptp_clock_ops);
 	unsigned long flags;
 
+	/* Setting systime not possible when using external TS clock. */
+	if (priv->plat->ext_sys_time)
+		return -EPERM;
+
 	spin_lock_irqsave(&priv->ptp_lock, flags);
 	stmmac_init_systime(priv, priv->ptpaddr, ts->tv_sec, ts->tv_nsec);
 	spin_unlock_irqrestore(&priv->ptp_lock, flags);
-- 
2.25.1

